server/auth/routes.js

```diff
@@ -7,48 +7,136 @@ const db = require("../services/database/users");
-		next(e);
-	}
+  const { email, password } = req.body;
+  console.log(`Login attempt ${email}`);
+  try {
+    const user = await db.getUserByEmail(email);
+
+    if (!email || !password || !user || password !== user.password) {
+      return res.sendStatus(403);
+    }
+
+    const secret = process.env.JWT_SECRET || "your_jwt_secret";
+    const token = jwt.sign({ userId: user.id }, secret);
+    delete user.password;
+    delete user.salt;
+    return res.send({ token, user });
+  } catch (e) {
+    console.error(e);
+    next(e);
+  }
 });
 
 /**
  * Users Registration
  */
 router.post("/register", async (req, res, next) => {
-	const { email, password } = req.body;
-
-	const user = {
-		email,
-		password
-	};
-
-	db.createUser(user)
-		.then(() => {
-			res.send({
-				success: true,
-				message: "Account created",
-			});
-		})
-		.catch((err) => {
-			console.log(err);
-			next(err);
-		});
+  const { email, password, name, role } = req.body;
+
+  const user = {
+    email,
+    password,
+    name,
+    role
+  };
+
+  //   if (
+  //     user.email != null &&
+  //     user.email.length > 1 &&
+  //     user.password != null &&
+  //     user.password.length > 1 &&
+  //     user.name != null &&
+  //     user.name.length > 1 &&
+  //     user.role != null &&
+  //     user.role.length > 1 &&
+  //     (user.role == "Interpreter" || user.role == "User")
+  //   ) {
+
+  //   } else {
+  //     res.status(400).send({
+  //       success: false,
+  //       message: "Empty Validation"
+  //     });
+  //   }
+
+  if (
+    validateEmailAddress(email) &&
+    validateName(name) &&
+    validatePassword(password) &&
+    validateRole(role)
+  ) {
+    db.createUser(user)
+      .then(() => {
+        res.send({
+          success: true,
+          message: "Account created"
+        });
+      })
+      .catch(err => {
+        res.status(400).send({
+          success: false,
+          message: err
+        });
+        console.log(err);
+        next(err);
+      });
+  } else {
+    res.send(err);
+  }
+  function validateEmailAddress(email) {
+    var expression = /(?!.*\.{2})^([a-z\d!#$%&'*+\-\/=?^_`{|}~\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]+(\.[a-z\d!#$%&'*+\-\/=?^_`{|}~\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]+)*|"((([ \t]*\r\n)?[ \t]+)?([\x01-\x08\x0b\x0c\x0e-\x1f\x7f\x21\x23-\x5b\x5d-\x7e\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]|\\[\x01-\x09\x0b\x0c\x0d-\x7f\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]))*(([ \t]*\r\n)?[ \t]+)?")@(([a-z\d\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]|[a-z\d\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF][a-z\d\-._~\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]*[a-z\d\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])\.)+([a-z\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]|[a-z\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF][a-z\d\-._~\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]*[a-z\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])\.?$/i;
+    if (expression.test(String(email).toLowerCase())) {
+      return true;
+    } else {
+      res.status(400).send({
```

Comment:
I would suggest to not use `res` in the validator function. A validator function can for example only return true or false with an error message that you use in the route implementation to return the right status and response.

Reply:
It was me who gave him an example of a validator function that also uses the res. I know it is not the best solution. Functions should stick to the Single Responsibility Principle.
What should be the best approach? return an object or passing a callback function?

Reply:
There are several approach but we could imagine having a validator function returning `true` if it passes or a string representing an error if it fails. Another approach could be to return systematically an object, `{result: true}` if it passes, `{result: false, error: "some message"}` if it fails. Moreover, if you define the validator functions in a different file, `res` will not be in scope anymore too :)

Reply:
I changed the validation error